Dockerize application

LeoVerto opened the pull request:
I use a multi-stage Dockerfile that builds the application in the first stage and copies over the binary to the second one
where only the necessary dependencies are installed.

Alpine has all the dependencies in it's main repositories, only log4cpp had to be installed from the testing one.

The resulting docker image is quite small at 117MB, a single-binary image would be even smaller but some of the libraries are not easily linked statically.

I had to make some changes to the cmake config, re-enable `registrarUri` for my SIP server and set the opus sample rate to the mumble default of 48000 or audio would be pitched down/up depending on the direction. I'm not sure if this is a regression or due to recent changes in mumlib.

If you don't want some of these changes in your main repo let me know and I'll modify the PR or feel free to cherry-pick whatever you want.

hiro2233 commented:
Hello, thanks for your pr.
I doubt that you have tested succefuly on Alpine natively, check my comments about this pr.

> I use a multi-stage Dockerfile that builds the application in the first stage and copies over the binary to the second one
> where only the necessary dependencies are installed.

That isn't the way, all dependencies need to be builded on the same box. On the docker script, add log4cpp build mechanism too.

> The resulting docker image is quite small at 117MB, a single-binary image would be even smaller but some of the libraries are not easily linked statically.

> I had to make some changes to the cmake config, re-enable registrarUri for my SIP server and set the opus sample rate to the mumble default of 48000 or audio would be pitched down/up depending on the direction. I'm not sure if this is a regression or due to recent changes in mumlib.

I fixed some bugs, make your changes, rebase this pr and build everything again.

hiro2233 commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+RUN mkdir mumsi-dist
+WORKDIR mumsi-dist
+
+RUN git clone https://github.com/hiro2233/mumlib.git \
```
now, this isn't necessary, remove cloning mumlib.

hiro2233 commented on MumbleCommunicator.cpp:
```diff
@@ -78,6 +78,7 @@ void mumble::MumbleCommunicator::connect(MumbleCommunicatorConfig &config) {
     mumConfig.opusEncoderBitrate = config.opusEncoderBitrate;
     mumConfig.cert_file = config.cert_file;
     mumConfig.privkey_file = config.privkey_file;
+    mumConfig.opusSampleRate = 48000;
```
remove it, mumlib is set to 16000 by default.

LeoVerto replied:
16000 didn't work for me and resulted in all audio received by other mumble clients to be massively downpitched and slowed down (as is to be expected with a sample rate mismatch) while telephony clients straight up didn't receive any audio (I suspect due to the codec cutting out extremely high frequencies) until I increased the samplerate.

LeoVerto replied:
Ah I see you've changed the sip sampling rate, I thought the sampling rate mismatch was between mumlib and the mumble server but I suppose this means it was between the two communicators.

hiro2233 replied:
build and test again without it, i made and pushed some fixed for this issue.

hiro2233 replied:
FYI, i performed all wire audio circuit on mumlib for 16Khz for 10 ms and lower latencies, but by default is set to 20 ms plus 1 size buffer, setting to 10 ms you will need to do some changes on your network nodes for beacons intervals, DTIMs intervals, Fragmentation Threshold, RTS Threshold and QoS settings to make it work.

LeoVerto replied:
Ah okay, that might explain why I had trouble getting 10ms to work.

hiro2233 commented on PjsuaCommunicator.cpp:
```diff
@@ -489,6 +489,8 @@ sip::PjsuaCommunicator::PjsuaCommunicator(IncomingConnectionValidator &validator
 
     endpoint.libInit(endpointConfig);
 
+    endpoint.audDevManager().setNullDev();
```
have you tried it with a mumble server?

LeoVerto replied:
Yes, I tested this and it worked fine.

As I understand it the current code already sets a null dev ([here](https://github.com/hiro2233/mumsi/blob/master/PjsuaCommunicator.cpp#L532)) but this change does the same thing but earlier, during initialization of the `PjsuaCommunicator` which prevents some warnings on startup due to Pjsua trying to select an audio device.

hiro2233 replied:
Test it again with the latest changes.

LeoVerto commented:
I've rebased on top of your changes, tested everything and added some documentation to the readme on how to build and run the docker version.

>  That isn't the way, all dependencies need to be builded on the same box. On the docker script, add log4cpp build mechanism too.

I'm not sure I understand what you mean by this, I use the log4cpp binary built by alpine maintainers but it isn't in the main repository yet so I have to pull it from the testing one.

As for the multi-stage build, the builder stage installs build tools and all the `-dev` versions of the packages and ends up being 560MB big while the runner stage only installs the packages required for running the app and comes in at 117MB.

hiro2233 commented:
> I'm not sure I understand what you mean by this, I use the log4cpp binary built by alpine maintainers but it isn't in the main repository yet so I have to pull it from the testing one.

ok, i didn't see [this](https://github.com/hiro2233/mumsi/pull/2/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R16), now it's ok.

hiro2233 commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+WORKDIR mumsi
+
+# Build mumlib from git submodule
+RUN mkdir -p modules/mumlib/build && cd modules/mumlib/build \
```
building mumlib is already present on the main mumsi cmake list, this isn't necessary.

hiro2233 replied:
https://github.com/hiro2233/mumsi/blob/master/CMakeLists.txt#L6
